File: android/app/src/main/java/com/prayertracker/app/data/ProfileRepository.kt
package com.prayertracker.app.data

import com.prayertracker.app.data.model.Profile
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Columns

class ProfileRepository(
    private val client: SupabaseClient = Supabase.client,
) {
    /** The signed-in user's profile, or null if not authenticated / not found. RLS scopes it. */
    suspend fun getMyProfile(): Profile? {
        val id = client.auth.currentUserOrNull()?.id ?: return null
        return client.from("profiles")
            .select(
                Columns.list(
                    "id", "display_name", "timezone", "latitude", "longitude",
                    "city_label", "calculation_method", "madhab",
                ),
            ) {
                filter { eq("id", id) }
            }
            .decodeSingleOrNull()
    }

    /** Updates only the columns the Settings screen owns. RLS restricts this to the caller's row. */
    suspend fun updateProfile(
        displayName: String,
        timezone: String,
        latitude: Double?,
        longitude: Double?,
        cityLabel: String?,
        calculationMethod: String,
        madhab: String,
    ) {
        val id = client.auth.currentUserOrNull()?.id ?: error("Not authenticated")
        client.from("profiles").update({
            set("display_name", displayName)
            set("timezone", timezone)
            set("latitude", latitude)
            set("longitude", longitude)
            set("city_label", cityLabel)
            set("calculation_method", calculationMethod)
            set("madhab", madhab)
        }) {
            filter { eq("id", id) }
        }
    }
}

File: web/src/data/profileRepository.ts
import { supabase } from '../lib/supabase'
import type { Profile } from './models'

const COLUMNS = 'id, display_name, timezone, latitude, longitude, city_label, calculation_method, madhab'

export async function getMyProfile(): Promise<Profile | null> {
  const {
    data: { user },
  } = await supabase.auth.getUser()
  if (!user) return null

  const { data, error } = await supabase.from('profiles').select(COLUMNS).eq('id', user.id).maybeSingle()
  if (error) throw error
  return data as Profile | null
}

export async function updateProfile(patch: {
  displayName: string
  timezone: string
  latitude: number | null
  longitude: number | null
  cityLabel: string | null
  calculationMethod: string
  madhab: string
}): Promise<void> {
  const {
    data: { user },
  } = await supabase.auth.getUser()
  if (!user) throw new Error('Not authenticated')

  const { error } = await supabase
    .from('profiles')
    .update({
      display_name: patch.displayName,
      timezone: patch.timezone,
      latitude: patch.latitude,
      longitude: patch.longitude,
      city_label: patch.cityLabel,
      calculation_method: patch.calculationMethod,
      madhab: patch.madhab,
    })
    .eq('id', user.id)
  if (error) throw error
}
